File: sukyung/200.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        visited = [[0] * n for _ in range(m)]  # land 방문 체크 배열
        dx = [1, -1, 0, 0]
        dy = [0, 0, 1, -1]
        island = 0

        def dfs(i, j):
            for k in range(4):
                new_i = i + dx[k]
                new_j = j + dy[k]
                # 아직 방문하지 않았고, 현재와 연결된 land인 경우
                # -> 하나의 섬으로 이어져 있는 경우임
                if 0<=new_i<m and 0<=new_j<n and not visited[new_i][new_j] and grid[new_i][new_j]=='1':
                    visited[new_i][new_j] = 1
                    dfs(new_i, new_j)  # 내 상하좌우로 연결된 land가 있는지 탐색


        for i in range(m):
            for j in range(n):
                # 아직 방문하지 않은 land이면 섬 검사
                if grid[i][j]=='1' and not visited[i][j]:
                    visited[i][j] == 1
                    dfs(i, j)
                    island += 1
        
        return island

File: dong99u/leetcode/200.java
package leetcode.leetcode200;

import java.util.*;

public class Solution {
	static int n, m; // 행 크기 m, 열 크기 n
	static int[] dxs = {1, 0, -1, 0};
	static int[] dys = {0, 1, 0, -1};

	public static int numIslands(char[][] grid) {
		m = grid.length;
		n = grid[0].length;

		boolean[][] visited = new boolean[m][n];

		int answer = 0;
		for (int i = 0; i < m; i++) {
			for (int j = 0; j < n; j++) {
				if (grid[i][j] == '1' &&!visited[i][j]) {
					dfs(i, j, grid, visited);
					answer++;
				}
			}
		}

		return answer;

    }

	static void dfs(int x, int y, char[][] grid, boolean[][] visited) {
		visited[x][y] = true;

		for (int i = 0; i < 4; i++) {
			int nx = x + dxs[i];
			int ny = y + dys[i];

			if (canGo(nx, ny, grid, visited)) {
				dfs(nx, ny, grid, visited);
			}

		}

	}

	static boolean canGo(int x, int y, char[][] grid, boolean[][] visited) {
		return inRange(x, y) && !visited[x][y] && grid[x][y] == '1';
	}

	static boolean inRange(int x, int y) {
		return (0 <= x && x < m) && (0 <= y && y < n);
	}
}
